pageserver/src/tenant/mgr.rs

```diff
@@ -774,20 +984,47 @@ where
         TenantsMap::Open(m) => m,
     };
     match m.entry(tenant_id) {
-        hash_map::Entry::Occupied(e) => Err(TenantMapInsertError::TenantAlreadyExists(
-            tenant_id,
-            e.get().current_state(),
-        )),
+        hash_map::Entry::Occupied(e) => match e.get() {
+            TenantSlot::Attached(t) => Err(TenantMapInsertError::TenantAlreadyExists(
+                tenant_id,
+                t.current_state(),
+            )),
+            TenantSlot::Secondary => Err(TenantMapInsertError::TenantExistsSecondary(tenant_id)),
+        },
         hash_map::Entry::Vacant(v) => match insert_fn().await {
             Ok(tenant) => {
-                v.insert(tenant.clone());
+                v.insert(TenantSlot::Attached(tenant.clone()));
                 Ok(tenant)
             }
             Err(e) => Err(TenantMapInsertError::Closure(e)),
         },
     }
 }
 
+async fn tenant_map_upsert_slot<'a, F, R>(
```

Comment:
Note to reviewers: this function is temporary.  `tenant_map_upsert_slot` and `tenant_map_insert` are both superseded by a more general mechanism in the next PR, which changes how concurrency control works for TenantsMap.